kimchi/src/tests/keccak.rs

```diff
@@ -1,54 +1,237 @@
+    print_matrix(&next[0..100]);
+}
+
+// Sets up test for a given message and desired input bytelength
+fn test_keccak<G: KimchiCurve>(message: BigUint) -> BigUint
+where
+    G::BaseField: PrimeField,
+{
+    let bytelength = message.to_bytes_be().len();
+    let padded_len = {
+        let mut sized = message.to_bytes_be();
+        sized.resize(bytelength - sized.len(), 0);
+        pad(&sized).len()
+    };
+    let _index = create_test_constraint_system::<G>(padded_len);
+    let witness = create_keccak_witness::<G>(message);
+
+    for r in 1..=24 {
+        print_witness::<G::ScalarField>(&witness, r);
+    }
+
+    let mut hash = vec![];
+    let hash_row = witness[0].len() - 2; // Hash row is dummy row
+    println!();
+    println!("----------------------------------------");
+    print!("Hash: ");
+    for b in 0..32 {
+        hash.push(FieldHelpers::to_bytes(&witness[200 + b][hash_row])[0]);
+        print!("{:02x}", hash[b]);
+    }
+    println!();
+    println!();
+
+    BigUint::from_bytes_be(&hash)
+}
+
+#[test]
+fn test_bitwise_sparse_representation() {
+    assert_eq!(expand(0xFFFF), 0x1111111111111111);
+
+    let word_a: u64 = 0x70d324ac9215fd8e;
+    let dense_a = decompose(word_a);
+    let real_dense_a = [0xfd8e, 0x9215, 0x24ac, 0x70d3];
+    for i in 0..QUARTERS {
+        assert_eq!(dense_a[i], real_dense_a[i]);
+    }
+    assert_eq!(word_a, compose(&dense_a));
+
+    let sparse_a = dense_a.iter().map(|x| expand(*x)).collect::<Vec<u64>>();
+    let real_sparse_a: Vec<u64> = vec![
+        0x1111110110001110,
+        0x1001001000010101,
+        0x10010010101100,
+        0x111000011010011,
+    ];
+    for i in 0..QUARTERS {
+        assert_eq!(sparse_a[i], real_sparse_a[i]);
+    }
+
+    let word_b: u64 = 0x11c76438a7f9e94d;
+    let dense_b = decompose(word_b);
+    let sparse_b = dense_b.iter().map(|x| expand(*x)).collect::<Vec<u64>>();
+
+    let xor_ab: u64 = word_a ^ word_b;
+    assert_eq!(xor_ab, 0x6114409435ec14c3);
+
+    let sparse_xor = decompose(xor_ab)
+        .iter()
+        .map(|x| expand(*x))
+        .collect::<Vec<u64>>();
+    let real_sparse_xor = [
+        0x1010011000011,
+        0x11010111101100,
+        0x100000010010100,
+        0x110000100010100,
+    ];
+    for i in 0..QUARTERS {
+        assert_eq!(sparse_xor[i], real_sparse_xor[i]);
+    }
+
+    let sparse_sum_ab = sparse_a
+        .iter()
+        .zip(sparse_b.iter())
+        .map(|(a, b)| a + b)
+        .collect::<Vec<u64>>();
+    let shifts_sum_ab = shift(&sparse_sum_ab);
+    let reset_sum_ab = reset(&shifts_sum_ab);
+    assert_eq!(sparse_xor, reset_sum_ab);
+
+    for i in 0..QUARTERS {
+        assert_eq!(
+            sparse_sum_ab[i],
+            shifts_sum_ab[i]
+                + shifts_sum_ab[4 + i] * 2
+                + shifts_sum_ab[8 + i] * 4
+                + shifts_sum_ab[12 + i] * 8
+        )
+    }
+}
+
+#[test]
+// Tests a random block of 1080 bits
+fn test_random_block() {
```

Comment:
This seems to be a test vector. Do you have a source we can check with?

Reply:
It is a random block of 1080 that I generated offline so I could hardcode the expected hash. You can check the value with this online tool for example: https://leventozturk.com/engineering/sha3/

If instead we want to be able to check the expected hash dynamically, we should make use of a library exposing an implementation of keccak.